src/internal/operators/takeWhile.ts

```diff
@@ -3,8 +3,8 @@ import { Observable } from '../Observable';
 import { Subscriber } from '../Subscriber';
 import { OperatorFunction, MonoTypeOperatorFunction, TeardownLogic } from '../types';
 
-export function takeWhile<T, S extends T>(predicate: (value: T, index: number) => value is S): OperatorFunction<T, S>;
-export function takeWhile<T>(predicate: (value: T, index: number) => boolean): MonoTypeOperatorFunction<T>;
+export function takeWhile<T, S extends T>(predicate: (value: T, index: number) => value is S, inclusive?: boolean): OperatorFunction<T, S>;
```

Comment:
Cool, but there'll need to be a few more signatures. The problem is that if `inclusive` can be `true`, the type guard has no meaning.

For example, consider a stream that is emitting carrots and potatoes and a type guard that accepts only carrots. If `inclusive` is `true`, the potato that completes the stream will be emitted, so the type guard should not narrow the emitted types. (Thanks go to @benlesh for the vegetable analogy.)

You can solve that problem by narrowing with a type guard only if `inclusive` is either omitted or is known to be `false`, like this:

```ts
export function takeWhile<T, S extends T>(predicate: (value: T, index: number) => value is S): OperatorFunction<T, S>;
export function takeWhile<T, S extends T>(predicate: (value: T, index: number) => value is S, inclusive: false): OperatorFunction<T, S>;
export function takeWhile<T>(predicate: (value: T, index: number) => boolean, inclusive: boolean): MonoTypeOperatorFunction<T>;
```

Thanks for the effort you've put into this PR.

Reply:
And if you could add another dtslint in which you explicitly pass `false` for `inclusive` - to make sure the signature with the `false` literal type returns a narrowed observable - that would be great.

Reply:
Makes sense. Thanks for the detailed explanation. I made the requested changes.